# Preventing Credit Card Defaults - Part 1

In this activity, you will use Keras to build a neural network model that predicts whether a credit card customer will default on their debt. You will apply what you learned in the previous demonstration. But, instead of using a dataset that contains 2 features, you will build a neural network model from a dataset that contains 22 features and 1 target.

## Instructions:

1. Read the CSV file from the Resources folder into a Pandas DataFrame.

2. Define the features set `X` by including all of the DataFrame columns except the “DEFAULT” column.

3. Create the target `y` by assigning the values of the DataFrame “DEFAULT” column.

4. Create the training and testing sets using the `train_test_split` function from scikit-learn.

5. Scale the features data using the `StandardScaler` from sklearn.

6. Create a neural network model with an input layer that consists of 22 inputs, one hidden layer, and an output layer. Use the `units` parameter to define 12 neurons for the hidden layer and a single output for the output layer. Use the ReLU activation function for the hidden layer and the sigmoid activation function for the output layer.

7. Display the model structure using the `summary` function.

## References:

[Keras Sequential model](https://keras.io/api/models/sequential/)

[Keras Dense module](https://keras.io/api/layers/core_layers/dense/)


In [2]:
# Imports
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

%matplotlib inline

## Step 1: Read the CSV file from the Resources folder into a Pandas DataFrame.

In [5]:
# Upload cc_default.csv to Colab
from google.colab import files

csv_file = files.upload()

Saving cc_default.csv to cc_default.csv


In [6]:
# Read the data into a Pandas DataFrame
cc_df = pd.read_csv("cc_default.csv")

# Review the DataFrame
cc_df.head()

LIMIT_BAL  EDUCATION  MARITAL_STATUS  ...  PAY_AMT5  PAY_AMT6  DEFAULT
0      20000          2               1  ...         0         0        1
1     120000          2               2  ...         0      2000        1
2      90000          2               2  ...      1000      5000        0
3      50000          2               1  ...      1069      1000        0
4      50000          2               1  ...       689       679        0

[5 rows x 23 columns]

## Step 2: Define the features set `X` by including all of the DataFrame columns except the “DEFAULT” column.

In [7]:
# Define features set X by selecting all columns but DEFAULT
X = cc_df.drop(columns=["DEFAULT"]).copy()

# Display the features DataFrame
X.head()

LIMIT_BAL  EDUCATION  MARITAL_STATUS  ...  PAY_AMT4  PAY_AMT5  PAY_AMT6
0      20000          2               1  ...         0         0         0
1     120000          2               2  ...      1000         0      2000
2      90000          2               2  ...      1000      1000      5000
3      50000          2               1  ...      1100      1069      1000
4      50000          2               1  ...      9000       689       679

[5 rows x 22 columns]

## Step 3: Create the target `y` by assigning the values of the DataFrame “DEFAULT” column.

In [8]:
# Define target set by selecting the DEFAULT column
y = cc_df["DEFAULT"]

# Display a sample y
y[:5]

0    1
1    1
2    0
3    0
4    0
Name: DEFAULT, dtype: int64

## Step 4: Create the training and testing sets using the `train_test_split` function from scikit-learn.

In [9]:
# Create training and testing datasets using train_test_split
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Step 5: Scale the features data using the `StandardScaler` from sklearn.

In [10]:
# Create the StandardScaler instance
X_scaler = StandardScaler()

In [11]:
# Fit the scaler to the features training dataset
X_scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [12]:
# Scale both the training and testing data from the features dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Step 6: Create a neural network model with an input layer that consists of 22 inputs, one hidden layer, and an output layer. Use the `units` parameter to define 12 neurons for the hidden layer and a single output for the output layer. Use the ReLU activation function for the hidden layer and the sigmoid activation function for the output layer.

In [13]:
# Imports
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential

In [14]:
# Define the the number of inputs to the model
number_inputs = 22

# Define the number of hidden nodes for the model
number_hidden_nodes = 12

# Create the Sequential model instance
neuron = Sequential()

# Add a Dense layer specifying the number of inputs, the number of hidden nodes, and the activation function
neuron.add(Dense(units=number_hidden_nodes, input_dim=number_inputs, activation="relu"))

# Add the output layer to the model specifying the number of output neurons and activation function
neuron.add(Dense(1, activation="sigmoid"))

## Step 7: Display the model structure using the `summary` function.

In [15]:
# Display the Sequential model summary
neuron.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 12)                276       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 13        
Total params: 289
Trainable params: 289
Non-trainable params: 0
_________________________________________________________________


##### The number of params for the first Dense layer is (22 x 12) + 12 = 276
##### The number of params for the second Dense layer is (12 x 1) + 1 = 13